File: src/agent/claude-agent/event-map.ts
import type { SDKMessage } from '@anthropic-ai/claude-agent-sdk';
import type { AgentEvent } from '../types';

/**
 * Map the Claude Agent SDK message stream to normalized {@link AgentEvent}s.
 *
 * Unlike codex's stateless `mapNotification`, the SDK stream needs a tiny bit of
 * per-turn state: it interleaves token-level `stream_event` deltas (only with
 * `includePartialMessages: true`) with whole `assistant`/`user` messages, and
 * text/thinking content blocks carry NO stable id. So we mint our own itemId per
 * content block and correlate the block's start → deltas → stop by its stream
 * `index` (an index is unique among currently-open blocks — blocks stream
 * sequentially within a message and a new message's block 0 only starts after the
 * previous message's blocks stopped). Tool calls DO carry a stable `id`, so those
 * map straight off the full `assistant`/`user` messages (no index needed).
 *
 * Observed message order (verified via spike, SDK 0.3.178):
 *   system/init → (per assistant API message) message_start →
 *     content_block_start(idx) → content_block_delta(idx)* →
 *     [full `assistant` message for that one block] → content_block_stop(idx) →
 *   … → [`user` message carrying tool_result blocks] → … → result
 * Each content block surfaces as its OWN `assistant` message (one block each).
 *
 * Event production per block:
 *   text     → text_delta(itemId)* (from deltas) then text(itemId) (at stop)
 *   thinking → thinking_delta(itemId)* then thinking(itemId)
 *   tool_use → tool_use(itemId = block.id) (from the assistant message)
 *   tool_result (user msg) → tool_result(itemId = tool_use_id)
 * The thread layer owns turn_started/done (synthetic turnId — the SDK has no turn
 * id; interrupt() needs none); this mapper emits everything else, and on `result`
 * emits usage + context_usage (the thread appends `done`).
 */

export interface ClaudeMapContext {
  /** thread cwd, to relativize file paths in tool titles */
  cwd?: string;
}

type Kind = 'text' | 'thinking' | 'tool_use' | 'other';
interface OpenBlock {
  itemId: string;
  kind: Kind;
  acc: string;
}

// Minimal structural shapes for the bits we read (avoids fighting the SDK's
// re-exported Anthropic Beta types; we only touch a handful of fields).
interface ContentBlock {
  type: string;
  id?: string;
  name?: string;
  input?: Record<string, unknown>;
  text?: string;
  thinking?: string;
  tool_use_id?: string;
  is_error?: boolean;
  content?: unknown;
}

export interface TurnMapper {
  /** Map one raw SDK message to zero or more AgentEvents (in order). */
  map(msg: SDKMessage): AgentEvent[];
}

/** Create a per-turn mapper. One per `runStreamed` turn (state resets with it). */
export function createTurnMapper(ctx: ClaudeMapContext = {}): TurnMapper {
  let blockSeq = 0;
  const open = new Map<number, OpenBlock>();
  let systemEmitted = false;
  // Captured from system/init — the result message has NO `model`, so the context
  // window fallback must read it here (else the window shows as 未知/unknown).
  let initModel: string | undefined;

  function map(msg: SDKMessage): AgentEvent[] {
    const m = msg as unknown as Record<string, any>;
    switch (m.type) {
      case 'system':
        if (m.subtype === 'init') {
          if (typeof m.model === 'string') initModel = m.model;
          if (!systemEmitted) {
            systemEmitted = true;
            return [{ type: 'system', threadId: String(m.session_id ?? '') }];
          }
        }
        // Auto-compaction during a turn → surface the「上下文已压缩」notice.
        if (m.subtype === 'compact_boundary') return [{ type: 'context_compacted' }];
        return [];

      case 'stream_event':
        return mapStreamEvent(m.event);

      case 'assistant': {
        const content = (m.message?.content ?? []) as ContentBlock[];
        const out: AgentEvent[] = [];
        for (const b of content) {
          if (b.type === 'tool_use' || b.type === 'server_tool_use') {
            out.push({
              type: 'tool_use',
              itemId: String(b.id ?? `tool${++blockSeq}`),
              title: toolTitle(b.name ?? '工具', b.input ?? {}, ctx.cwd),
              detail: toolDetail(b.name ?? '', b.input ?? {}),
            });
          }
        }
        return out;
      }

      case 'user': {
        const content = m.message?.content;
        if (!Array.isArray(content)) return [];
        const out: AgentEvent[] = [];
        for (const b of content as ContentBlock[]) {
          if (b.type === 'tool_result') {
            out.push({
              type: 'tool_result',
              itemId: String(b.tool_use_id ?? ''),
              output: toolResultText(b.content),
              exitCode: b.is_error ? 1 : 0,
            });
          }
        }
        return out;
      }

      case 'result': {
        // The turn boundary. We emit only usage/context here; the THREAD inspects
        // the result message to decide `done` vs `error` (it knows whether the
        // user pressed ⏹ — an interrupted turn ends with subtype
        // 'error_during_execution', which is NOT a real failure). See thread.ts.
        const out: AgentEvent[] = [];
        const usage = m.usage as Record<string, number> | undefined;
        if (usage) {
          out.push({
            type: 'usage',
            inputTokens: usage.input_tokens,
            outputTokens: usage.output_tokens,
          });
          // Fallback context gauge (the thread overrides this with the authoritative
          // getContextUsage() reading). used = input + cache_read + cache_creation
          // (≈ the request's full input context — verified to equal getContextUsage
          // totalTokens). Window comes from the init model, NOT result.model (absent).
          const used =
            (usage.input_tokens ?? 0) +
            (usage.cache_read_input_tokens ?? 0) +
            (usage.cache_creation_input_tokens ?? 0);
          if (used > 0) {
            out.push({ type: 'context_usage', usedTokens: used, contextWindow: contextWindowFor(initModel) });
          }
        }
        return out;
      }

      // Transient API hiccup the SDK auto-retries — surface a retrying footer
      // (run-state keeps the ⏹ button and lets the next deltas overwrite it).
      case 'system_api_retry':
      case 'api_retry':
        return [{ type: 'error', message: '网络波动，正在重试…', willRetry: true }];

      default:
        return [];
    }
  }

  function mapStreamEvent(ev: Record<string, any> | undefined): AgentEvent[] {
    if (!ev || typeof ev.type !== 'string') return [];
    switch (ev.type) {
      case 'content_block_start': {
        const idx = ev.index as number;
        const cb = (ev.content_block ?? {}) as ContentBlock;
        const kind: Kind =
          cb.type === 'thinking' || cb.type === 'redacted_thinking'
            ? 'thinking'
            : cb.type === 'text'
              ? 'text'
              : cb.type === 'tool_use' || cb.type === 'server_tool_use'
                ? 'tool_use'
                : 'other';
        open.set(idx, { itemId: `b${++blockSeq}`, kind, acc: '' });
        return [];
      }
      case 'content_block_delta': {
        const b = open.get(ev.index as number);
        if (!b) return [];
        const d = (ev.delta ?? {}) as Record<string, any>;
        if (d.type === 'text_delta' && typeof d.text === 'string') {
          b.acc += d.text;
          return [{ type: 'text_delta', itemId: b.itemId, delta: d.text }];
        }
        if (d.type === 'thinking_delta' && typeof d.thinking === 'string') {
          b.acc += d.thinking;
          return [{ type: 'thinking_delta', itemId: b.itemId, delta: d.thinking }];
        }
        return [];
      }
      case 'content_block_stop': {
        const idx = ev.index as number;
        const b = open.get(idx);
        open.delete(idx);
        if (!b || !b.acc) return [];
        if (b.kind === 'text') return [{ type: 'text', itemId: b.itemId, text: b.acc }];
        if (b.kind === 'thinking') return [{ type: 'thinking', itemId: b.itemId, text: b.acc }];
        return [];
      }
      default:
        return [];
    }
  }

  return { map };
}

// ── tool rendering helpers ──────────────────────────────────────────────────

/** A Claude built-in tool call → a short Chinese title, mirroring codex's
 * command-as-title convention (Bash shows the command itself). Exported so the
 * history folder (resume card) renders tool calls identically to the live stream. */
export function toolTitle(name: string, input: Record<string, unknown>, cwd?: string): string {
  const s = (k: string): string => (typeof input[k] === 'string' ? (input[k] as string) : '');
  switch (name) {
    case 'Bash':
    case 'BashOutput':
      return s('command') || 'Shell 命令';
    case 'Read':
      return `读取 ${displayPath(s('file_path'), cwd)}`;
    case 'Write':
      return `写入 ${displayPath(s('file_path'), cwd)}`;
    case 'Edit':
    case 'MultiEdit':
      return `编辑 ${displayPath(s('file_path'), cwd)}`;
    case 'NotebookEdit':
      return `编辑笔记本 ${displayPath(s('notebook_path'), cwd)}`;
    case 'Glob':
      return `查找 ${s('pattern')}`.trim();
    case 'Grep':
      return `搜索 ${s('pattern')}`.trim();
    case 'WebFetch':
      return `抓取网页 ${s('url')}`.trim();
    case 'WebSearch':
      return `联网搜索 ${s('query')}`.trim();
    case 'Task':
      return `子任务：${s('description') || s('subagent_type') || ''}`.trim();
    case 'TodoWrite':
      return '更新待办清单';
    case 'ExitPlanMode':
      return '提交方案';
    default:
      return name || '工具调用';
  }
}

/** Secondary line for a tool block (e.g. a Bash command's description). */
function toolDetail(name: string, input: Record<string, unknown>): string | undefined {
  if (name === 'Bash' && typeof input.description === 'string' && input.description) {
    return input.description;
  }
  return undefined;
}

/** tool_result.content is a string or an array of {type:'text',text} (and image)
 * blocks — flatten to plain text for the card's tool output panel. */
function toolResultText(content: unknown): string | undefined {
  if (content == null) return undefined;
  if (typeof content === 'string') return content || undefined;
  if (Array.isArray(content)) {
    const text = content
      .map((b) => {
        const x = b as Record<string, any>;
        if (x?.type === 'text' && typeof x.text === 'string') return x.text;
        if (x?.type === 'image') return '[图片]';
        return '';
      })
      .join('')
      .trim();
    return text || undefined;
  }
  return undefined;
}

const PATH_TAIL_MAX = 40;
/** Relative path inside cwd / absolute outside it (so an out-of-project touch is
 * visible); with no cwd, long paths keep only trailing segments. Mirrors
 * codex event-map's displayPath. */
function displayPath(p: string, cwd?: string): string {
  if (!p) return '文件';
  if (cwd) {
    const sep = cwd.includes('\\') ? '\\' : '/';
    const root = cwd.endsWith(sep) ? cwd : cwd + sep;
    if (p.startsWith(root) && p.length > root.length) return p.slice(root.length);
  }
  if (p.length <= PATH_TAIL_MAX || !p.includes('/')) return p;
  const segs = p.split('/');
  let out = segs[segs.length - 1] ?? p;
  for (let i = segs.length - 2; i >= 0; i--) {
    const cand = `${segs[i]}/${out}`;
    if (cand.length > PATH_TAIL_MAX) break;
    out = cand;
  }
  return `…/${out}`;
}

/** Model context window heuristic — the SDK's result message doesn't report it.
 * `[1m]`-suffixed / 1m-context models get 1,000,000; otherwise the 200k default.
 * Approximate (drives only the gauge percentage); refine if the SDK exposes it. */
function contextWindowFor(model: unknown): number | null {
  const id = typeof model === 'string' ? model.toLowerCase() : '';
  if (!id) return null;
  if (id.includes('1m') || id.includes('[1m]')) return 1_000_000;
  return 200_000;
}

/** Human-readable error text for a non-success `result` message. Used by the
 * thread when a turn ends in failure (not an interrupt). */
export function resultErrorText(m: Record<string, unknown>): string {
  const result = m.result;
  if (typeof result === 'string' && result.trim()) return result.trim();
  const subtype = m.subtype;
  if (typeof subtype === 'string') {
    if (subtype === 'error_max_turns') return '已达到最大轮次限制';
    if (subtype === 'error_max_budget_usd') return '已达到预算上限';
    return `运行出错（${subtype}）`;
  }
  return '运行出错';
}

File: internal/agent/claude/eventmap.go
package claude

// eventmap.go —— claude stream-json 消息 → 归一化 AgentEvent（纯函数，港口 TS event-map.ts）。
//
// 与 codex 的无状态 mapNotification 不同，Claude 的 SDK 流需要一点 per-turn 状态：
// 它把 token 级 stream_event 增量（需 includePartialMessages）与完整的 assistant/user
// 消息交错；text/thinking 内容块没有稳定 id。于是我们为每个内容块自造 itemId，并用流的
// index 关联「块开始 → 增量 → 结束」。工具调用带稳定 id，直接从完整消息映射。
//
// 观测到的消息顺序（对齐 TS 注释）：
//   system/init → [每轮 API 消息] message_start → content_block_start(idx) →
//     content_block_delta(idx)* → [该块自身的完整 assistant 消息] → content_block_stop(idx)
//   → … → [携带 tool_result 的 user 消息] → … → result
// 每个内容块作为它【自己的】assistant 消息出现（每块一个）。
//
// 每块的事件产出：
//   text     → text_delta(itemId)*（增量）然后 text(itemId)（结束）
//   thinking → thinking_delta(itemId)* 然后 thinking(itemId)
//   tool_use → tool_use(itemId = block.id)（来自 assistant 消息）
//   tool_result（user 消息）→ tool_result(itemId = tool_use_id)
// turn 的 turn_started/done 由 thread 层合成（SDK 无 turn id）；本 mapper 产出其余一切，
// 并在 result 时产出 usage + context_usage（thread 追加上 done）。

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
)

// ── 原始消息解析 ──────────────────────────────────────────────

type claudeMessage struct {
	Type      string          `json:"type"`
	Subtype   string          `json:"subtype"`
	SessionID string          `json:"session_id"`
	Model     string          `json:"model"`
	Message   json.RawMessage `json:"message"` // assistant/user
	Event     json.RawMessage `json:"event"`   // stream_event
	Usage     json.RawMessage `json:"usage"`   // result
	Result    string          `json:"result"`  // result.subtype 的附带文本
}

func parseClaudeMessage(raw json.RawMessage) claudeMessage {
	var m claudeMessage
	_ = json.Unmarshal(raw, &m)
	return m
}

// ── 内容块 / 增量形状 ───────────────────────────────────────────

type contentBlock struct {
	Type       string         `json:"type"`
	ID         string         `json:"id"`
	Name       string         `json:"name"`
	Input      map[string]any `json:"input"`
	Text       string         `json:"text"`
	Thinking   string         `json:"thinking"`
	ToolUseID  string         `json:"tool_use_id"`
	IsError    bool           `json:"is_error"`
	Content    json.RawMessage `json:"content"`
}

type contentDelta struct {
	Type     string `json:"type"`
	Text     string `json:"text"`
	Thinking string `json:"thinking"`
}

// ── per-turn mapper ──────────────────────────────────────────

type openBlock struct {
	itemID string
	kind   string // text | thinking | tool_use | other
	acc    string
}

// turnMapper 一个 runStreamed/runGoal turn 一个实例（状态随 turn 重置）。
type turnMapper struct {
	blockSeq      int
	open          map[int]*openBlock
	systemEmitted bool
	initModel     string
	sawPartial    bool // 本 turn 是否见过 stream_event 增量（用于 text/thinking 兜底）
	cwd           string
}

func createTurnMapper(cwd string) *turnMapper {
	return &turnMapper{open: map[int]*openBlock{}, cwd: cwd}
}

// mapMsg 把一条原始消息映射为零到多个 AgentEvent（按顺序）。
func (m *turnMapper) mapMsg(msg claudeMessage) []agent.AgentEvent {
	switch msg.Type {
	case "system":
		if msg.Subtype == "init" {
			if msg.Model != "" {
				m.initModel = msg.Model
			}
			if !m.systemEmitted {
				m.systemEmitted = true
				return []agent.AgentEvent{agent.EvSys(msg.SessionID)}
			}
		}
		// turn 内自动压缩 → 弹出「上下文已压缩」提示。
		if msg.Subtype == "compact_boundary" {
			return []agent.AgentEvent{agent.EvCompacted()}
		}
		return nil

	case "stream_event":
		return m.mapStreamEvent(msg.Event)

	case "assistant":
		return m.mapAssistant(msg.Message)

	case "user":
		return m.mapUser(msg.Message)

	case "result":
		return m.mapResult(msg)

	case "system_api_retry", "api_retry":
		// 瞬态 API 抖动（SDK 自动重试）→ 弹出「重试中」脚注。
		return []agent.AgentEvent{agent.EvErrorT("网络波动，正在重试…", true)}
	}
	return nil
}

func (m *turnMapper) mapStreamEvent(event json.RawMessage) []agent.AgentEvent {
	if len(event) == 0 {
		return nil
	}
	var ev struct {
		Type         string       `json:"type"`
		Index        int          `json:"index"`
		ContentBlock contentBlock `json:"content_block"`
		Delta        contentDelta `json:"delta"`
	}
	if json.Unmarshal(event, &ev) != nil {
		return nil
	}
	switch ev.Type {
	case "content_block_start":
		m.sawPartial = true
		kind := "other"
		switch ev.ContentBlock.Type {
		case "thinking", "redacted_thinking":
			kind = "thinking"
		case "text":
			kind = "text"
		case "tool_use", "server_tool_use":
			kind = "tool_use"
		}
		m.blockSeq++
		m.open[ev.Index] = &openBlock{itemID: fmt.Sprintf("b%d", m.blockSeq), kind: kind}
		return nil

	case "content_block_delta":
		b := m.open[ev.Index]
		if b == nil {
			return nil
		}
		if ev.Delta.Type == "text_delta" && ev.Delta.Text != "" {
			b.acc += ev.Delta.Text
			return []agent.AgentEvent{agent.EvTextD(b.itemID, ev.Delta.Text)}
		}
		if ev.Delta.Type == "thinking_delta" && ev.Delta.Thinking != "" {
			b.acc += ev.Delta.Thinking
			return []agent.AgentEvent{agent.EvThinkingD(b.itemID, ev.Delta.Thinking)}
		}
		return nil

	case "content_block_stop":
		b := m.open[ev.Index]
		delete(m.open, ev.Index)
		if b == nil || b.acc == "" {
			return nil
		}
		if b.kind == "text" {
			return []agent.AgentEvent{agent.EvTextFull(b.itemID, b.acc)}
		}
		if b.kind == "thinking" {
			return []agent.AgentEvent{agent.EvThinkingFull(b.itemID, b.acc)}
		}
		return nil
	}
	return nil
}

func (m *turnMapper) mapAssistant(message json.RawMessage) []agent.AgentEvent {
	if len(message) == 0 {
		return nil
	}
	var msg struct {
		Content []contentBlock `json:"content"`
	}
	if json.Unmarshal(message, &msg) != nil {
		return nil
	}
	out := []agent.AgentEvent{}
	for _, b := range msg.Content {
		if b.Type == "tool_use" || b.Type == "server_tool_use" {
			out = append(out, agent.EvToolUK(b.ID, toolTitle(b.Name, b.Input, m.cwd), toolDetail(b.Name, b.Input), toolKind(b.Name)))
			continue
		}
		// text/thinking：仅当本 turn 没见过 stream_event 增量（sawPartial=false）
		// 时才从完整 assistant 消息补发——增量模式下它们已由 content_block_stop 发出。
		if m.sawPartial {
			continue
		}
		if b.Type == "text" && b.Text != "" {
			m.blockSeq++
			out = append(out, agent.EvTextFull(fmt.Sprintf("b%d", m.blockSeq), b.Text))
		} else if (b.Type == "thinking" || b.Type == "redacted_thinking") && b.Thinking != "" {
			m.blockSeq++
			out = append(out, agent.EvThinkingFull(fmt.Sprintf("b%d", m.blockSeq), b.Thinking))
		}
	}
	return out
}

func (m *turnMapper) mapUser(message json.RawMessage) []agent.AgentEvent {
	if len(message) == 0 {
		return nil
	}
	var msg struct {
		Content []contentBlock `json:"content"`
	}
	if json.Unmarshal(message, &msg) != nil {
		return nil
	}
	out := []agent.AgentEvent{}
	for _, b := range msg.Content {
		if b.Type == "tool_result" {
			ec := 0
			if b.IsError {
				ec = 1
			}
			out = append(out, agent.EvToolR(b.ToolUseID, toolResultText(b.Content), &ec))
		}
	}
	return out
}

func (m *turnMapper) mapResult(msg claudeMessage) []agent.AgentEvent {
	out := []agent.AgentEvent{}
	if len(msg.Usage) == 0 {
		return out
	}
	var u struct {
		InputTokens    int `json:"input_tokens"`
		OutputTokens   int `json:"output_tokens"`
		CacheRead      int `json:"cache_read_input_tokens"`
		CacheCreation  int `json:"cache_creation_input_tokens"`
	}
	if json.Unmarshal(msg.Usage, &u) != nil {
		return out
	}
	out = append(out, agent.EvUsageT(u.InputTokens, u.OutputTokens))
	// 兜底上下文量（thread 用权威读数覆盖）；used = input + cache_read + cache_creation。
	used := u.InputTokens + u.CacheRead + u.CacheCreation
	if used > 0 {
		out = append(out, agent.EvContext(used, contextWindowFor(m.initModel)))
	}
	return out
}

// ── 工具渲染助手 ──────────────────────────────────────────────

// toolTitle 把 Claude 内置工具调用渲染成简短中文标题（对齐 codex 的「命令即标题」约定）。
func toolTitle(name string, input map[string]any, cwd string) string {
	s := func(k string) string {
		if v, ok := input[k]; ok {
			if str, ok := v.(string); ok {
				return str
			}
		}
		return ""
	}
	switch name {
	case "Bash", "BashOutput":
		return orDefault(s("command"), "Shell 命令")
	case "Read":
		return "读取 " + displayPath(s("file_path"), cwd)
	case "Write":
		return "写入 " + displayPath(s("file_path"), cwd)
	case "Edit", "MultiEdit":
		return "编辑 " + displayPath(s("file_path"), cwd)
	case "NotebookEdit":
		return "编辑笔记本 " + displayPath(s("notebook_path"), cwd)
	case "Glob":
		return "查找 " + s("pattern")
	case "Grep":
		return "搜索 " + s("pattern")
	case "WebFetch":
		return "抓取网页 " + s("url")
	case "WebSearch":
		return "联网搜索 " + s("query")
	case "Task":
		return "子任务：" + orDefault(s("description"), s("subagent_type"))
	case "TodoWrite":
		return "更新待办清单"
	case "ExitPlanMode":
		return "提交方案"
	default:
		return orDefault(name, "工具调用")
	}
}

// toolDetail 工具块的副标题（如 Bash 命令的 description）。
func toolDetail(name string, input map[string]any) string {
	if name == "Bash" {
		if d, ok := input["description"].(string); ok && d != "" {
			return d
		}
	}
	return ""
}

// toolKind 工具的粗分类（对齐 TS event-map.ts toolKind），供卡片按类别渲染：
// command=Bash 类（标题即完整命令，渲染为 bash 代码块）；file=文件读/写/编辑；search=grep/glob/web；
// 其余 → tool。与 toolTitle 的 switch 保持一致。
func toolKind(name string) agent.ToolKind {
	switch name {
	case "Bash", "BashOutput":
		return agent.ToolKindCommand
	case "Read", "Write", "Edit", "MultiEdit", "NotebookEdit":
		return agent.ToolKindFile
	case "Glob", "Grep", "WebFetch", "WebSearch":
		return agent.ToolKindSearch
	default:
		return agent.ToolKindTool
	}
}

// toolResultText tool_result.content 是 string 或 {type:'text',text} 数组（含图片）→ 纯文本。
func toolResultText(content json.RawMessage) string {
	if len(content) == 0 {
		return ""
	}
	var raw any
	if json.Unmarshal(content, &raw) != nil {
		return ""
	}
	switch v := raw.(type) {
	case string:
		return strings.TrimSpace(v)
	case []any:
		var sb strings.Builder
		for _, b := range v {
			obj, ok := b.(map[string]any)
			if !ok {
				continue
			}
			if obj["type"] == "text" {
				if t, ok := obj["text"].(string); ok {
					sb.WriteString(t)
				}
			} else if obj["type"] == "image" {
				sb.WriteString("[图片]")
			}
		}
		return strings.TrimSpace(sb.String())
	}
	return ""
}

const pathTailMax = 40

// displayPath cwd 内相对化；cwd 外保留绝对（让越项目改动可见）；无 cwd 时长路径只留尾部段。
func displayPath(p, cwd string) string {
	if p == "" {
		return "文件"
	}
	if cwd != "" {
		sep := "/"
		if strings.Contains(cwd, "\\") {
			sep = "\\"
		}
		root := cwd
		if !strings.HasSuffix(root, sep) {
			root += sep
		}
		if strings.HasPrefix(p, root) && len(p) > len(root) {
			return p[len(root):]
		}
		return p
	}
	if len(p) <= pathTailMax || !strings.Contains(p, "/") {
		return p
	}
	segs := strings.Split(p, "/")
	out := segs[len(segs)-1]
	for i := len(segs) - 2; i >= 0; i-- {
		cand := segs[i] + "/" + out
		if len(cand) > pathTailMax {
			break
		}
		out = cand
	}
	return "…/" + out
}

// contextWindowFor 模型上下文窗口启发式——result 消息不报窗口。
// [1m] 后缀 / 1m 上下文模型 → 1,000,000；否则（含空模型）200k 默认。仅驱动仪表盘百分比。
func contextWindowFor(model string) *int {
	id := strings.ToLower(model)
	var w int
	if strings.Contains(id, "1m") || strings.Contains(id, "[1m]") {
		w = 1_000_000
	} else {
		w = 200_000
	}
	return &w
}

// resultErrorText 非成功 result 消息的人类可读错误（turn 失败时 thread 使用）。
func resultErrorText(m claudeMessage) string {
	if strings.TrimSpace(m.Result) != "" {
		return strings.TrimSpace(m.Result)
	}
	switch m.Subtype {
	case "error_max_turns":
		return "已达到最大轮次限制"
	case "error_max_budget_usd":
		return "已达到预算上限"
	}
	if m.Subtype != "" {
		return "运行出错（" + m.Subtype + "）"
	}
	return "运行出错"
}

func orDefault(s, def string) string {
	if strings.TrimSpace(s) != "" {
		return s
	}
	return def
}
